### ISSUE TITLE
Release 0.25 - April 2019

### ISSUE THREAD

**Issue by laurentlb** [bazelbuild/bazel#7498]
Target RC date: April 1st

**Comment by dslomov** [comment]
We would like to get fixes for #6426 (/cc @aragos) and #7910 (/cc @buchgr) into 0.25. Let's make an ETA for a cut is April 3rd - if the fixes are not available by then, we will cut anyway - ok with you @dkelmer ?

**Comment by dkelmer** [comment]
SGTM

I would also like to mention that the fix for https://github.com/bazelbuild/bazel/issues/7773 should probably be in the release as the behavior is gated by an incompatible change flag so people that migrated post the 0.24 release where the flag became available will now be broken.

**Comment by dslomov** [comment]
Update #6426 is not making it :(

**Comment by buchgr** [comment]
Update #7910 has been fixed.

**Comment by dkelmer** [comment]
Great! The latest green of the Bazel@HEAD + Downstream is from before those commits so I will trigger a new run now and then cut from that

**Comment by dkelmer** [comment]
Waiting for this run to complete: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/898#c5fe1ee3-fbfb-4222-bcda-a6647b64e147

**Comment by dkelmer** [comment]
To create first RC!
`scripts/release/release.sh create 0.25.0 03662462941953dad23af88693804b8b1d4430b9`

**Comment by dkelmer** [comment]
Downstream test of rc1: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/901

**Comment by laszlocsomor** [comment]
Can I request a cherrypick to flip an incompatible flag (https://github.com/bazelbuild/bazel/issues/7486)?

**Comment by laszlocsomor** [comment]
> Can I request a cherrypick to flip an incompatible flag (#7486)?

If yes, please cherry-pick afeb8d0b7fef619159fc8fbaaeb8bd41dd2619bd. Thanks!

**Comment by laszlocsomor** [comment]
Another potential cherry-pick:
- 0.25.0 will be a migration window for `--incompatible_windows_native_test_wrapper` (https://github.com/bazelbuild/bazel/issues/6622)
- I just found a bug in this feature (https://github.com/bazelbuild/bazel/issues/7956), that I can fix easily
- is it fine to cherry-pick this bugfix, so I don't have to guard the bugfix with another incompatible flag?

**Comment by dkelmer** [comment]
@laszlocsomor: Regarding #7486, did you run a "tgp" with the flag flip? [Instructions here](https://github.com/bazelbuild/continuous-integration/blob/master/docs/downstream-testing.md). If it's clean I'll cherry pick it. 

And I will cherry pick the bug fix for #7956. Ping this issue when that is ready and I'll cherry pick both together (assuming the flag flip is clean-ish)

**Comment by laszlocsomor** [comment]
@dkelmer : Yes and it [all passed](https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/903), see https://github.com/bazelbuild/bazel/pull/7646#issuecomment-480215177. 

Great, I'll do as you asked. Thanks!

**Comment by laszlocsomor** [comment]
Merged fix for #7956. Please cherrypick https://github.com/bazelbuild/bazel/commit/3f7f255f9a8345b8898e3953e7e53d68106cc63d.

**Comment by dkelmer** [comment]
To create 0.25rc2
`scripts/release/release.sh create --force_rc=2 0.25.0 03662462941953dad23af88693804b8b1d4430b9 3f7f255f9a8345b8898e3953e7e53d68106cc63d afeb8d0b7fef619159fc8fbaaeb8bd41dd2619bd`

**Comment by buchgr** [comment]
https://github.com/bazelbuild/bazel/pull/8008 is a release blocker. I ll ping this thread once it's fixed.

**Comment by limdor** [comment]
Some findings of breaking changes in bazel 0.25 + cc_library + action_env + USERNAME variable, not enough information yet to open a proper bug:
https://groups.google.com/d/msg/bazel-dev/zotC5HGla8I/y3rgU8wjCwAJ

**Comment by buchgr** [comment]
#8008 has been fixed. Please cherry-pick 4299b6549cbc1b3e4494c91ed2f51d49b14c7980

**Comment by dkelmer** [comment]
Command to create rc3:
`scripts/release/release.sh create --force_rc=3 0.25.0 03662462941953dad23af88693804b8b1d4430b9 3f7f255f9a8345b8898e3953e7e53d68106cc63d afeb8d0b7fef619159fc8fbaaeb8bd41dd2619bd 4299b6549cbc1b3e4494c91ed2f51d49b14c7980`

**Comment by tuananhckpt90** [comment]
A

**Comment by tuananhckpt90** [comment]
A

**Comment by tuananhckpt90** [comment]
A

**Comment by tuananhckpt90** [comment]
Một

**Comment by dkelmer** [comment]
@limdor Any update on the thread/bug?

**Comment by limdor** [comment]
For the moment I know that is related to:
disk cache + cc_library + .a files + action_env=USERNAME + Bazel from 0.24
to 0.25 rc
The problem is that I could still not get a minimum reproduceable. On
Wednesday I will try again if I can get one.

**Comment by limdor** [comment]
@dkelmer I tried to reproduce it with a clean disk cache and I could not get a reproducible even with my original project. Unfortunately I didn't keep a copy of the cache to investigate deeper.

**Comment by steeve** [comment]
Hey guys, is it possible to cherry-pick https://github.com/bazelbuild/bazel/commit/0020a97fdc20ca099ec6386771b20d3236f9890d ? 

Not a regression per se but Xcode 10.2 breaks without it.

**Comment by meteorcloudy** [comment]
@dkelmer  We need to cherry-pick https://github.com/bazelbuild/bazel/commit/130f86ded1ce84f959f0b78c065211902faed546 to fix https://github.com/bazelbuild/bazel/issues/8104

**Comment by dkelmer** [comment]
`scripts/release/release.sh create --force_rc=8 0.25.0 03662462941953dad23af88693804b8b1d4430b9 3f7f255f9a8345b8898e3953e7e53d68106cc63d afeb8d0b7fef619159fc8fbaaeb8bd41dd2619bd 4299b6549cbc1b3e4494c91ed2f51d49b14c7980 231270c67d5aa771462245531fa9b2ee7d3d0ae8 75a3a531b08e727ade4fa3cb0a574bd142727cce 4a6354a3a5ca23583f8b62e3e439a04ce75b863f ae102fbde3c1ff87e4f67007a275fb30792a4e8d 0020a97fdc20ca099ec6386771b20d3236f9890d 130f86ded1ce84f959f0b78c065211902faed546`

**Comment by dkelmer** [comment]
@steeve @meteorcloudy, cherrypicking now

**Comment by keith** [comment]
@steeve @dkelmer that fix is already in 0.25?

**Comment by dkelmer** [comment]
@keith, you're right. Thanks for catching that. I think it doesn't matter that it will be listed as a cherry pick though so I will proceed with rc8 which has @meteorcloudy's cherry pick in it.

**Comment by hlopko** [comment]
It seems https://github.com/bazelbuild/bazel/issues/8104 was introduced in 0.23 (was it @meteorcloudy?), so it's not a 0.24 regression strictly speaking (if I remember the policy correctly :) Couldn't the fix wait for 0.26 which we're cutting this Thursday? Delaying 0.25 means we will not be able to test incompatible changes before 0.26 is cut.

**Comment by dkelmer** [comment]
@hlopko, Ah, sorry I didn't notice that and already pushed the RC. So question 1 would be: is it even possible to release an older RC? Even if it is, question 2 is: why do you need 0.25 to be out to test incompatible changes? I thought you could run a custom presubmit with your flags? I will note that I'm going to release on Wednesday morning NY time, so you would have a few hours to test with incompatible changes before 0.26 gets cut.

**Comment by dkelmer** [comment]
Release blocking: https://buganizer.corp.google.com/issues/131462201

**Comment by hlopko** [comment]
Wednesday is a public holiday in Germany :) I'm not concerned by any specific incompatible flag I have, rather about the team velocity and happy users. It is true that if there was a flag I cared about, I could test it manually. But in that case I cannot migrate downstream project proactively - because they still use 0.24, and my behavior is only in on 0.25.

When 0.25 is out, our users can start testing incompatible flags and report back if the migration is not possible or contains bugs. Then we have the option to extend the migration window and fix the issue in 0.26. Without this feedback we have to either flip the flag blind, or wait for multiple releases.

We also don't have time for users to try 0.25, report regressions, and for us to fix regressions before 0.26 is cut. This will result in more cherrypicks in 0.26, which can slow down 0.27 and so on :)

**Comment by steeve** [comment]
> @steeve @dkelmer that fix is already in 0.25?

My bad, I though RCs were tagged too, and since I didn't see it in a tag I wrongly assumed it wasn't. I'm glad it is !

**Comment by dkelmer** [comment]
@hlopko 
I mostly agree with your points. It's definitely important for people to test a particular release before we cut a new one. But I think that ideally users would try out the RC and report regressions with the RC so that we can fix the current RC, as opposed to add cherrypicks to the next release. However I do realize that users tend to not do this and I don't know the solution to that.  

In the case of this particular release, I don't think 24 hours would've made a big difference but I am sorry nonetheless that it will be so tight :(

**Comment by Kernald** [comment]
Just a small note on this 0.25.0 release: something is slightly off with the changelog.

> `--incompatible_disable_legacy_crosstool_fields` has been flipped (#6861)

is repeated three times. The line just below the second occurrence seems incomplete:
<img width="548" alt="Screen Shot 2019-05-02 at 10 56 13 AM" src="https://user-images.githubusercontent.com/296840/57052165-f5867700-6cc8-11e9-9f97-108e13f706fa.png">

**Comment by shahms** [comment]
It would appear as though this release removed previously working functionality in cc_common.compile/link without any replacement (not just an incompatible API).  This breaks a chunk of tests for Kythe.

**Comment by shahms** [comment]
Because of this, 0.25 is incompatible with Kythe, full-stop.  It's one thing to change an interface in a backwards-incompatible way and entirely different to remove the implementation entirely when doing so.

**Comment by dkelmer** [comment]
Hi @shahms, I'm cc'ing @hlopko who will be more informed. Do you have some examples of breakages to help narrow down the issue?

**Comment by dkelmer** [comment]
Bazel 0.25 is released. You can see it here: https://github.com/bazelbuild/bazel/releases

@vbatts, @petemounce, and @excitoon Would you be able to update the various package definitions? Thanks!!

**Comment by shahms** [comment]
https://github.com/bazelbuild/bazel/issues/8226 is a bug specifically about this issue, but essentially 976876d changed the cc_common.compile API in a backwards compatible way (fine) but also simply removed the old implementation of that API so that they previously working (if suboptimal) functions now just return None.  This isn't something we can work around.

**Comment by dkelmer** [comment]
Thanks for narrowing it down and filing the bug. cc'ing @oquenchil who was the author of the cited change to weigh in on what to do. @shahms, is it a problem for you to hold off on upgrading to 0.25 until tomorrow when we figure out what to do about the change?

**Comment by shahms** [comment]
In that I will have to manually downgrade to 0.24 because my employer pushes new releases of Bazel to workstations automatically, it's annoying but manageable ;-) Thanks for looking into it.

**Comment by vbatts** [comment]
fedora and centos builds are up.
Not a blocker, but build failed for rawhide with:
```
third_party/grpc/src/core/lib/gpr/log_linux.cc:43:13: error: ambiguating new declaration of 'long int gettid()'
   43 | static long gettid(void) { return syscall(__NR_gettid); }
      |             ^~~~~~
In file included from /usr/include/unistd.h:1170,
                 from third_party/grpc/src/core/lib/gpr/log_linux.cc:41:
```
That is with gcc 9.0.1-0.15.fc31

**Comment by vbatts** [comment]
https://copr.fedorainfracloud.org/coprs/vbatts/bazel/builds/ for more information. Specifically builder-live.log from https://copr-be.cloud.fedoraproject.org/results/vbatts/bazel/fedora-rawhide-x86_64/00902055-bazel/

**Comment by excitoon** [comment]
lukesampson/scoop#3433

**Comment by jin** [comment]
0.25 broke rules_scala. Same pipeline worked on 0.24.

https://buildkite.com/bazel/rules-jvm-external-examples/builds/259#f5b93b74-ed72-4217-b392-e339912a72a0

```
ERROR: /home/bazel/.cache/bazel/_bazel_bazel/600edb71338d22ba878d955ee9e0e953/external/io_bazel_rules_scala/third_party/unused_dependency_checker/src/main/BUILD:5:1: in scala_library_for_plugin_bootstrapping rule @io_bazel_rules_scala//third_party/unused_dependency_checker/src/main:unused_dependency_checker:
--
  | Traceback (most recent call last):
  | File "/home/bazel/.cache/bazel/_bazel_bazel/600edb71338d22ba878d955ee9e0e953/external/io_bazel_rules_scala/third_party/unused_dependency_checker/src/main/BUILD", line 5
  | scala_library_for_plugin_bootstrapping(name = 'unused_dependency_checker')
  | File "/home/bazel/.cache/bazel/_bazel_bazel/600edb71338d22ba878d955ee9e0e953/external/io_bazel_rules_scala/scala/private/rule_impls.bzl", line 769, in scala_library_for_plugin_bootstrapping_impl
  | _lib(ctx, scalac_provider.default_class..., <3 more arguments>)
  | File "/home/bazel/.cache/bazel/_bazel_bazel/600edb71338d22ba878d955ee9e0e953/external/io_bazel_rules_scala/scala/private/rule_impls.bzl", line 689, in _lib
  | _compile_or_empty(ctx, ctx.outputs.manifest, cjars, srcj..., <6 more arguments>)
  | File "/home/bazel/.cache/bazel/_bazel_bazel/600edb71338d22ba878d955ee9e0e953/external/io_bazel_rules_scala/scala/private/rule_impls.bzl", line 467, in _compile_or_empty
  | java_common.run_ijar(ctx.actions, jar = ctx.outputs.jar, <2 more arguments>)
  | expected value of type 'JavaToolchainSkylarkApiProvider' for parameter 'java_toolchain', for call to method run_ijar(actions, jar, target_label = None, java_toolchain) of 'java_common'
```

cc @iirina

**Comment by petemounce** [comment]
Published to chocolatey

**Comment by dkelmer** [comment]
@iirina can you take a look at https://github.com/bazelbuild/bazel/issues/7498#issuecomment-489286392? Though @jin, I'm a little surprised this is the case because it didn't show up in the test with downstream projects: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/939

**Comment by dkelmer** [comment]
Update on the 0.25 release: We're doing a patch release to address https://github.com/bazelbuild/bazel/issues/8226. The fix will be to provide an implementation and the API that will be available in 0.26.

**Comment by laurentlb** [comment]
Did you send an announcement to the blog and bazel-discuss@?

**Comment by dkelmer** [comment]
@laurentlb, no. I can send an email to bazel-discuss@. What do you mean by sending an announcement to the blog?

**Comment by jin** [comment]
@dkelmer @iirina you can ignore https://github.com/bazelbuild/bazel/issues/7498#issuecomment-489286392 - fixed by bumping rules_scala.

**Comment by dkelmer** [comment]
fix to https://github.com/bazelbuild/bazel/issues/8236 also needs to be included in the patch release

**Comment by dkelmer** [comment]
Command to create 0.25.1rc3
`scripts/release/release.sh create --force_rc=3 0.25.1 f44bc4732650f6ad011407cc5c41cb36de23acdb ed48a4a5fddbd93b057c3aa726e15720d79dcf8f 81aefe7ee01cc73646a53f9c72ed40ead09f9f5a`

Note that first commit is the 0.25.0 release, the second is a local commit, and the 3rd is a public commit

**Comment by dkelmer** [comment]
Patch release 0.25.1 has been released https://github.com/bazelbuild/bazel/releases/tag/0.25.1

**Comment by petemounce** [comment]
Pushed to chocolatey.

**Comment by jin** [comment]
The Android emulator team recently pushed a new stable emulator release that removed some files that Bazel's Android SDK BUILD template looks for (emulator64-x86 and emulator64-arm). Because of this, *all* Bazel versions will fail when trying to test `android_instrumentation_test` targets or run `android_device` rules. The problem is that there is no way for users to download an older version of the emulator from Android Studio or the `sdkmanager` command line, so users who update their Android emulators will see this breakage with no workaround.

https://github.com/bazelbuild/bazel/issues/8280

This affects all Bazel versions, and has already been [cherry-picked into 0.26.0rc7](https://github.com/bazelbuild/bazel/issues/7499#issuecomment-491200810). Can we cherry-pick the fix to 0.25 too? The fix is here: https://github.com/bazelbuild/bazel/commit/85a5a2bd569a5274950fc7327a044c395248c024

cc @meteorcloudy

**Comment by dkelmer** [comment]
This is an interesting case because it's not a Bazel breakage, but I do think that it makes sense for the most recent version of Bazel to be able to build an `android_device` so I'll create a patch release (we can't cherry pick PRs into an existing release). But to clarify: there's no way the emulator auto updates, right? And so is it not possible that some people _will_ still have an older version of the emulator image? Once https://github.com/bazelbuild/bazel/commit/85a5a2bd569a5274950fc7327a044c395248c024 is cherry picked, won't that mean that the patch release won't work with those older emulators?

**Comment by jin** [comment]
The emulator will not automatically update, but new SDK downloads will use the new emulator. The patch is also backwards compatible with older emulator versions.

**Comment by dkelmer** [comment]
Backwards compatible is what I wanted to hear ;) 
Starting the patch release process now

**Comment by dkelmer** [comment]
Also adding the fix to https://github.com/bazelbuild/bazel/issues/8254 to the patch release

**Comment by dkelmer** [comment]
To create 0.25.2:

`scripts/release/release.sh create --force_rc=2 0.25.2 582b79ef3e2c9c9ae1878eedd2771df11c5f6061 0900660d67b53a56a13d1fa16a788e4cecbb1c0e 85a5a2bd569a5274950fc7327a044c395248c024`

**Comment by buchgr** [comment]
The fix for this will require a patch release, as the bug effectively causes an an infinite loop for remote execution: https://github.com/bazelbuild/bazel/issues/8320

**Comment by dkelmer** [comment]
Created 0.25.3
`scripts/release/release.sh create 0.25.3 318aebb523b3c91275855e35f0a67814e0d2231f 0ff19c6d0adf3c0df94fff59ca3bd13cbcf99897`

**Comment by petemounce** [comment]
I tried to create the 0.25.3 chocolatey package, but https://github.com/bazelbuild/bazel/releases/tag/0.25.3 does not exist yet?

**Comment by dkelmer** [comment]
@petemounce, that is correct. I'm pushing it now. There was an issue that has been resolved.
